### PR TITLE
OCPCLOUD-2016: Surface cpms vs machine diff

### PR DESCRIPTION
**Pull request by damdo**
To more easily understand what causes machine replacements triggered by the Control Plane Machine Set, we can surface the ProviderConfig (a.k.a. ProviderSpec) difference that is found between the CPMS (desired) ProviderSpec and the Machine's (current) ProviderSpec.

### CODE REVIEW

File: pkg/controllers/controlplanemachineset/updates.go [hunk 1]
```diff
@@ -424,7 +424,7 @@ func (r *ControlPlaneMachineSetReconciler) createRollingUpdateReplacementMachine
 		// Trigger a Machine creation.
 		// Consider the first found outdated machine for this index to be the one in need of update.
 		outdatedMachine := machinesNeedingReplacement[0]
-		logger := logger.WithValues("index", outdatedMachine.Index, "namespace", r.Namespace, "name", outdatedMachine.MachineRef.ObjectMeta.Name)
+		logger := logger.WithValues("index", outdatedMachine.Index, "namespace", r.Namespace, "name", outdatedMachine.MachineRef.ObjectMeta.Name, "diff", outdatedMachine.Diff)
```
Comment (wking) [inline pkg/controllers/controlplanemachineset/updates.go:427]: Also possibly something similar for the `OnDelete` strategy?  Which would maybe be around [here][1]

[1]: https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/850c514d3874ac5595252bc7b672c2db3ccc37be/pkg/controllers/controlplanemachineset/updates.go#L375-L392

Comment (damdo) [reply pkg/controllers/controlplanemachineset/updates.go:427]: Yes yes 👍  I was planning to add that too but wanted to test things out first.
Added now.

File: pkg/test/resourcebuilder/machineproviders/machine_info.go [hunk 2]
```diff
@@ -81,9 +83,19 @@ func (m MachineInfoBuilder) Build() machineproviders.MachineInfo {
...
+	if !m.needsUpdate && m.diff != nil {
+		panic("There shall not be Diff if NeedsUpdate is false")
+	}
+
 	return info
 }
 
+// WithDiff sets the needsupdate for the machineinfo builder.
+func (m MachineInfoBuilder) WithDiff(diff []string) MachineInfoBuilder {
+	m.diff = diff
```
Comment (wking) [inline pkg/test/resourcebuilder/machineproviders/machine_info.go:95]: did you want to set `needsUpdate = true` here too?

Comment (damdo) [reply pkg/test/resourcebuilder/machineproviders/machine_info.go:99]: Added

**Comment by wking** [comment]
Hrm, [two][1] [runs][2] both seem to have had CPMS scale-up issues.  I'm not clear on the mechanism.  Digging into [azure-operator][1], it's recognizing the need for an update:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616/artifacts/e2e-azure-operator/gather-extra/artifacts/controlplanemachinesets.json | jq -r '.items[].status.conditions[] | .lastTransitionTime + " " + .type + "=" + .status + " " + .reason + ": " + .message'
2023-03-22T17:26:13Z Error=False AsExpected: 
2023-03-22T17:28:00Z Available=True AllReplicasAvailable: 
2023-03-22T17:26:49Z Degraded=False AsExpected: 
2023-03-22T17:49:24Z Progressing=True NeedsUpdateReplicas: Observed 1 replica(s) in need of update
```

Checking in [the operator logs][3]:

```
2023-03-22T17:49:24.817466431Z I0322 17:49:24.817429       1 status.go:119]  "msg"="Observed Machine Configuration" "controller"="controlplanemachineset" "name"="cluster" "namespace"="openshift-machine-api" "observedGeneration"=1 "readyReplicas"=3 "reconcileID"="24602126-b63c-4533-b6eb-d0438e223a73" "replicas"=3 "unavailableReplicas"=0 "updatedReplicas"=2
2023-03-22T17:49:24.891861214Z I0322 17:49:24.891761       1 provider.go:464]  "msg"="Created machine" "controller"="controlplanemachineset" "diff"=["VMSize: Standard_D8s_v3 != Standard_D16s_v3"] "failureDomain"="AzureFailureDomain{Zone:2}" "index"=1 "machineName"="ci-op-jtsk7ngh-f5128-r25k2-master-vbntn-1" "name"="ci-op-jtsk7ngh-f5128-r25k2-master-1" "namespace"="openshift-machine-api" "reconcileID"="24602126-b63c-4533-b6eb-d0438e223a73" "updateStrategy"="RollingUpdate"
2023-03-22T17:49:24.891861214Z I0322 17:49:24.891813       1 updates.go:478]  "msg"="Created replacement machine" "controller"="controlplanemachineset" "diff"=["VMSize: Standard_D8s_v3 != Standard_D16s_v3"] "index"=1 "name"="ci-op-jtsk7ngh-f5128-r25k2-master-1" "namespace"="openshift-machine-api" "reconcileID"="24602126-b63c-4533-b6eb-d0438e223a73" "updateStrategy"="RollingUpdate"
2023-03-22T17:49:24.891940015Z I0322 17:49:24.891855       1 controller.go:118]  "msg"="Reconciling control plane machine set" "controller"="controlplanemachinesetgenerator" "name"="cluster" "namespace"="openshift-machine-api" "reconcileID"="98589f4c-f95d-41e0-ae65-013693ea4263"
```

So hooray, the `diff` business looks great :).  But ~5 minutes later, while the operator is still waiting on the new Machine, the logs stop:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616/artifacts/e2e-azure-operator/gather-must-gather/artifacts/must-gather/inspect.local.6265849191083317081/namespaces/openshift-machine-api/pods/control-plane-machine-set-operator-cb8d54bf-dgdf2/control-plane-machine-set-operator/control-plane-machine-set-operator/logs/current.log | tail -n2
2023-03-22T17:54:02.279035475Z I0322 17:54:02.278987       1 updates.go:287]  "msg"="Waiting for replacement machine to become ready" "controller"="controlplanemachineset" "index"=1 "name"="ci-op-jtsk7ngh-f5128-r25k2-master-1" "namespace"="openshift-machine-api" "reconcileID"="64575441-b79e-4eb0-8314-b1e11577c6e1" "replacementName"="ci-op-jtsk7ngh-f5128-r25k2-master-vbntn-1" "updateStrategy"="RollingUpdate"
2023-03-22T17:54:02.279910189Z I0322 17:54:02.279876       1 controller.go:208]  "msg"="Finished reconciling control plane machine set" "controller"="controlplanemachineset" "name"="cluster" "namespace"="openshift-machine-api" "reconcileID"="64575441-b79e-4eb0-8314-b1e11577c6e1"
```

This is despite the cousin Machine controller container continuing to log over 30m later:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616/artifacts/e2e-azure-operator/gather-must-gather/artifacts/must-gather/inspect.local.6265849191083317081/namespaces/openshift-machine-api/pods/machine-api-controllers-9dc7b6786-sjwrh/machine-controller/machine-controller/logs/current.log | tail -n1
2023-03-22T18:35:19.951418149Z I0322 18:35:19.951406       1 machine_scope.go:216] ci-op-jtsk7ngh-f5128-r25k2-master-1: patching machine
```

And [the Machine looks healthy][5].

Smells like a CPMS-controller goroutine wedged, although I'm not sure what the detailed mechanism would be.  I guess the fact that we have similar symptoms in two CI runs sounds promising for being able to reproduce, and maybe we can [SIGQUIT the operator after reproducing to force a stack dump][4].

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-ovn-etcd-scaling/1638583875196162048
[3]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616/artifacts/e2e-azure-operator/gather-must-gather/artifacts/must-gather/inspect.local.6265849191083317081/namespaces/openshift-machine-api/pods/control-plane-machine-set-operator-cb8d54bf-dgdf2/control-plane-machine-set-operator/control-plane-machine-set-operator/logs/current.log
[4]: https://pkg.go.dev/os/signal#hdr-Default_behavior_of_signals_in_Go_programs
[5]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616/artifacts/e2e-azure-operator/gather-must-gather/artifacts/must-gather/inspect.local.6265849191083317081/namespaces/openshift-machine-api/machine.openshift.io/machines/ci-op-jtsk7ngh-f5128-r25k2-master-vbntn-1.yaml

**Comment by wking** [comment]
Wildly guessing:

```code
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616/artifacts/e2e-azure-operator/gather-must-gather/artifacts/must-gather/inspect.local.6265849191083317081/namespaces/openshift-machine-api/pods/control-plane-machine-set-operator-cb8d54bf-dgdf2/control-plane-machine-set-operator/control-plane-machine-set-operator/logs/current.log | grep 'Observed Machine Configuration' | tail -n1
2023-03-22T17:54:02.278606167Z I0322 17:54:02.278549       1 status.go:119]  "msg"="Observed Machine Configuration" "controller"="controlplanemachineset" "name"="cluster" "namespace"="openshift-machine-api" "observedGeneration"=1 "readyReplicas"=3 "reconcileID"="64575441-b79e-4eb0-8314-b1e11577c6e1" "replicas"=4 "unavailableReplicas"=0 "updatedReplicas"=2
```

looks like it could be the transition to 3 `updatedReplicas` as the new node goes ready which triggered the lock?

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638583875162607616/artifacts/e2e-azure-operator/gather-extra/artifacts/nodes.json | jq -r '.items[] | select(.metadata.annotations["machine.openshift.io/machine"] == "openshift-machine-api/ci-op-jtsk7ngh-f5128-r25k2-master-vbntn-1").status.conditions[] | select(.type == "Ready") | .lastTransitionTime + " "  + .lastHeartbeatTime + " " + .type + "=" + .status + " " + .reason + ": " + .message'
2023-03-22T17:54:01Z 2023-03-22T18:35:00Z Ready=False KubeletNotReady: container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/. Has your network provider started?
```

Hmm, 17:54:01 for the Node's `Ready` `lastTransitionTime` is suspiciously close to the CPMS-controller's 17:54:02 final log entry.  But the Node is also only `Ready=False`, so there is more going wrong than just the CPMS controller apparently wedging.

**Comment by wking** [comment]
[Recent azure-operator run][1] shows #177 having fixed the hang, with similar timestamps for the CPMS operator:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856/artifacts/e2e-azure-operator/gather-must-gather/artifacts/must-gather/inspect.local.730974301449140309/namespaces/openshift-machine-api/pods/control-plane-machine-set-operator-678df5cbbd-9pl2t/control-plane-machine-set-operator/control-plane-machine-set-operator/logs/current.log | tail -n3
2023-03-23T15:57:12.026369939Z I0323 15:57:12.026348       1 status.go:119]  "msg"="Observed Machine Configuration" "controller"="controlplanemachineset" "name"="cluster" "namespace"="openshift-machine-api" "observedGeneration"=1 "readyReplicas"=3 "reconcileID"="2531f5ed-3af9-403e-8f19-d9cdd072cbb5" "replicas"=4 "unavailableReplicas"=0 "updatedReplicas"=2
2023-03-23T15:57:12.026612640Z I0323 15:57:12.026594       1 updates.go:316]  "msg"="Waiting for replacement machine to become ready" "controller"="controlplanemachineset" "index"=1 "name"="ci-op-mcr9zhfk-f5128-dg8pj-master-1" "namespace"="openshift-machine-api" "reconcileID"="2531f5ed-3af9-403e-8f19-d9cdd072cbb5" "replacementName"="ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1" "updateStrategy"="RollingUpdate"
2023-03-23T15:57:12.027075141Z I0323 15:57:12.027057       1 controller.go:208]  "msg"="Finished reconciling control plane machine set" "controller"="controlplanemachineset" "name"="cluster" "namespace"="openshift-machine-api" "reconcileID"="2531f5ed-3af9-403e-8f19-d9cdd072cbb5"
```

and the Machine controller:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856/artifacts/e2e-azure-operator/gather-must-gather/artifacts/must-gather/inspect.local.730974301449140309/namespaces/openshift-machine-api/pods/machine-api-controllers-54ffdfd444-8mhhq/machine-controller/machine-controller/logs/current.log | tail -n1
2023-03-23T15:56:01.065296381Z I0323 15:56:01.065209       1 machine_scope.go:216] ci-op-mcr9zhfk-f5128-dg8pj-worker-centralus3-jrkch: patching machine
```

But it's still not happy with `ci-op-mcr9zhfk-f5128-dg8pj-master-1`, and it's not clear on why.  While we're fiddling with debugability in logging, do we want to expand `Waiting for replacement machine to become ready` to include an `error` or similar string that explains why we don't think it's ready yet?  Things look happy to me:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856/artifacts/e2e-azure-operator/gather-extra/artifacts/machines.json | jq -r '.items[] | select(.metadata.name == "ci-op-mcr9zhfk-f5128-dg8pj-master-1").status.phase'
Running
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856/artifacts/e2e-azure-operator/gather-extra/artifacts/nodes.json | jq -r '.items[] | select(.metadata.name == "ci-op-mcr9zhfk-f5128-dg8pj-master-1").status.conditions[] | select(.type == "Ready")'
{
  "lastHeartbeatTime": "2023-03-23T15:54:16Z",
  "lastTransitionTime": "2023-03-23T14:55:28Z",
  "message": "kubelet is posting ready status",
  "reason": "KubeletReady",
  "status": "True",
  "type": "Ready"
}
```

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856

**Comment by damdo** [comment]
@wking yes #177 fixed the hang.
I believe the issue we are still seeing is the [other Azure](https://issues.redhat.com/browse/OCPBUGS-7359) bug [I mentioned above](https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/180#issuecomment-1480766613).

From the kubelet logs we can see the new Node doesn't have outbound connection:
```
[~] $ curl -SsL https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856/artifacts/e2e-azure-operator/gather-extra/artifacts/nodes/ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1/journal | zcat | tail -n 10

Mar 23 15:57:56.385111 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.384767    2300 pod_workers.go:965] "Error syncing pod, skipping" err="network is not ready: container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/. Has your network provider started?" pod="openshift-etcd/revision-pruner-8-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1" podUID=3ef5b79b-384f-4d1f-acc5-b005de5ada76
Mar 23 15:57:56.385111 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.384831    2300 pod_workers.go:965] "Error syncing pod, skipping" err="network is not ready: container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/. Has your network provider started?" pod="openshift-e2e-loki/loki-promtail-lk4kw" podUID=952b4a5c-6e4e-423b-85ae-42b345ace3ad
Mar 23 15:57:56.385111 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.384903    2300 pod_workers.go:965] "Error syncing pod, skipping" err="network is not ready: container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/. Has your network provider started?" pod="openshift-kube-apiserver/revision-pruner-7-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1" podUID=9e1b470d-bc23-4c10-8abf-f90e37d68b2e
Mar 23 15:57:56.391364 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 crio[2268]: time="2023-03-23 15:57:56.391314089Z" level=info msg="runSandbox: releasing container name: k8s_POD_azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager_a0d4c7c9-3dde-4f74-94fd-a675872fa2dc_0" id=978e8069-83f2-452d-af29-845d4b099f45 name=/runtime.v1.RuntimeService/RunPodSandbox
Mar 23 15:57:56.391799 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 crio[2268]: time="2023-03-23 15:57:56.391368590Z" level=info msg="runSandbox: releasing pod sandbox name: k8s_azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager_a0d4c7c9-3dde-4f74-94fd-a675872fa2dc_0" id=978e8069-83f2-452d-af29-845d4b099f45 name=/runtime.v1.RuntimeService/RunPodSandbox
Mar 23 15:57:56.391835 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.391541    2300 remote_runtime.go:176] "RunPodSandbox from runtime service failed" err="rpc error: code = DeadlineExceeded desc = creating pod sandbox with name \"k8s_azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager_a0d4c7c9-3dde-4f74-94fd-a675872fa2dc_0\": initializing source docker://registry.build05.ci.openshift.org/ci-op-mcr9zhfk/stable@sha256:4dceac41db909978cfbbfd308a3d9aae121530ae288bd82dc0583a9e6553fc9d: pinging container registry registry.build05.ci.openshift.org: Get \"https://registry.build05.ci.openshift.org/v2/\": dial tcp 54.165.2.8:443: i/o timeout"
Mar 23 15:57:56.391835 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.391583    2300 kuberuntime_sandbox.go:72] "Failed to create sandbox for pod" err="rpc error: code = DeadlineExceeded desc = creating pod sandbox with name \"k8s_azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager_a0d4c7c9-3dde-4f74-94fd-a675872fa2dc_0\": initializing source docker://registry.build05.ci.openshift.org/ci-op-mcr9zhfk/stable@sha256:4dceac41db909978cfbbfd308a3d9aae121530ae288bd82dc0583a9e6553fc9d: pinging container registry registry.build05.ci.openshift.org: Get \"https://registry.build05.ci.openshift.org/v2/\": dial tcp 54.165.2.8:443: i/o timeout" pod="openshift-cloud-controller-manager/azure-cloud-node-manager-tw8cx"
Mar 23 15:57:56.391835 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.391607    2300 kuberuntime_manager.go:782] "CreatePodSandbox for pod failed" err="rpc error: code = DeadlineExceeded desc = creating pod sandbox with name \"k8s_azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager_a0d4c7c9-3dde-4f74-94fd-a675872fa2dc_0\": initializing source docker://registry.build05.ci.openshift.org/ci-op-mcr9zhfk/stable@sha256:4dceac41db909978cfbbfd308a3d9aae121530ae288bd82dc0583a9e6553fc9d: pinging container registry registry.build05.ci.openshift.org: Get \"https://registry.build05.ci.openshift.org/v2/\": dial tcp 54.165.2.8:443: i/o timeout" pod="openshift-cloud-controller-manager/azure-cloud-node-manager-tw8cx"
Mar 23 15:57:56.391835 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.391658    2300 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"CreatePodSandbox\" for \"azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager(a0d4c7c9-3dde-4f74-94fd-a675872fa2dc)\" with CreatePodSandboxError: \"Failed to create sandbox for pod \\\"azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager(a0d4c7c9-3dde-4f74-94fd-a675872fa2dc)\\\": rpc error: code = DeadlineExceeded desc = creating pod sandbox with name \\\"k8s_azure-cloud-node-manager-tw8cx_openshift-cloud-controller-manager_a0d4c7c9-3dde-4f74-94fd-a675872fa2dc_0\\\": initializing source docker://registry.build05.ci.openshift.org/ci-op-mcr9zhfk/stable@sha256:4dceac41db909978cfbbfd308a3d9aae121530ae288bd82dc0583a9e6553fc9d: pinging container registry registry.build05.ci.openshift.org: Get \\\"https://registry.build05.ci.openshift.org/v2/\\\": dial tcp 54.165.2.8:443: i/o timeout\"" pod="openshift-cloud-controller-manager/azure-cloud-node-manager-tw8cx" podUID=a0d4c7c9-3dde-4f74-94fd-a675872fa2dc
Mar 23 15:57:56.968246 ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1 kubenswrapper[2300]: E0323 15:57:56.968207    2300 kubelet.go:2495] "Container runtime network not ready" networkReady="NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: No CNI configuration file in /etc/kubernetes/cni/net.d/. Has your network provider started?"
```

And all the pods scheduled on it are in `Pending` state (stuck in `ContainerCreating`/`PodInitializing`):
```
[~] $ curl -SsL https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856/artifacts/e2e-azure-operator/gather-extra/artifacts/pods.json | jq -r '.items[] | select(.spec.nodeName == "ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1") | .status.phase +" "+ .metadata.name'

Pending azure-cloud-node-manager-tw8cx
Pending azure-disk-csi-driver-node-d6trr
Pending azure-file-csi-driver-node-k7kpx
Pending tuned-wb5t6
Pending node-resolver-6x6d9
Pending loki-promtail-lk4kw
Pending installer-9-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending revision-pruner-8-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending revision-pruner-9-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending node-ca-w24b8
Pending apiserver-watcher-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending installer-7-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending revision-pruner-7-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending installer-6-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending revision-pruner-6-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending installer-6-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending revision-pruner-6-ci-op-mcr9zhfk-f5128-dg8pj-master-npcf8-1
Pending machine-config-daemon-ttvqr
Pending node-exporter-2ggj2
Pending multus-additional-cni-plugins-4gfh6
Pending multus-pvbxn
Pending network-metrics-daemon-6j78n
Pending network-check-target-cc9wf
Pending ovnkube-master-rmntq
Pending ovnkube-node-96ghz
```

This leads to the new master Node not being able to host a new etcd member, preventing the machine deletion hook to delete the old, outdated, outgoing master `ci-op-mcr9zhfk-f5128-dg8pj-master-1` (see Drainable condition below).

```
[~] $ curl -SsL https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-control-plane-machine-set-operator/180/pull-ci-openshift-cluster-control-plane-machine-set-operator-main-e2e-azure-operator/1638905660319993856/artifacts/e2e-azure-operator/gather-extra/artifacts/machines.json | jq -r '.items[] | select(.metadata.name == "ci-op-mcr9zhfk-f5128-dg8pj-master-1") | .status'
{
  "addresses": [
    {
      "address": "ci-op-mcr9zhfk-f5128-dg8pj-master-1",
      "type": "Hostname"
    },
    {
      "address": "ci-op-mcr9zhfk-f5128-dg8pj-master-1",
      "type": "InternalDNS"
    },
    {
      "address": "ci-op-mcr9zhfk-f5128-dg8pj-master-1.43vuytjh0bjezga1nw2qr2d1ze.gx.internal.cloudapp.net",
      "type": "InternalDNS"
    },
    {
      "address": "10.0.0.8",
      "type": "InternalIP"
    }
  ],
  "conditions": [
    {
      "lastTransitionTime": "2023-03-23T14:53:26Z",
      "message": "Drain operation currently blocked by: [{Name:EtcdQuorumOperator Owner:clusteroperator/etcd}]",
      "reason": "HookPresent",
      "severity": "Warning",
      "status": "False",
      "type": "Drainable"
    },
    {
      "lastTransitionTime": "2023-03-23T14:52:17Z",
      "status": "True",
      "type": "InstanceExists"
    },
    {
      "lastTransitionTime": "2023-03-23T14:51:18Z",
      "status": "True",
      "type": "Terminable"
    }
  ],
  "lastUpdated": "2023-03-23T14:53:26Z",
  "nodeRef": {
    "kind": "Node",
    "name": "ci-op-mcr9zhfk-f5128-dg8pj-master-1",
    "uid": "f5d8f8c3-c898-42d6-80f9-22548ccfa1f1"
  },
  "phase": "Running",
  "providerStatus": {
    "conditions": [
      {
        "lastTransitionTime": "2023-03-23T14:52:07Z",
        "message": "machine successfully created",
        "reason": "MachineCreationSucceeded",
        "status": "True",
        "type": "MachineCreated"
      }
    ],
    "metadata": {},
    "vmId": "/subscriptions/d38f1e38-4bed-438e-b227-833f997adf6a/resourceGroups/ci-op-mcr9zhfk-f5128-dg8pj-rg/providers/Microsoft.Compute/virtualMachines/ci-op-mcr9zhfk-f5128-dg8pj-master-1",
    "vmState": "Running"
  }
}
```

Which is exactly the symptom of wedged master we are seeing on Azure in OCPBUGS-7359.

The reason why the new node/machine comes up as Ready/Running as you pointed out in [your previous message](https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/180#issuecomment-1483100995) still remains a mystery to me. Given that the CPMS is detecting either one or the other condition as Un-Ready or Not in Running phase (could it be caching?).

**Comment by JoelSpeed** [comment]
So one note here is that the code doesn't actually log the diff when it's detected unless the log level is set to `4` or, there is an error in the process. So, I think we might want to adjust the logging here to achieve our goal.

I think we probably want to print some information in the gathering stage to say, we found this machine needing update and print the diff. That said, `needsUpdate` isn't only tied to the diff right? We have some rules around replacing if the machine is deleted?

I would suggest we find a place to put a log line at level 2 that says "Machine need replacement", with a reason of "Machine is terminating" or "Machine is out of date" and when it's out of date, include the diff

WDYT?

**Comment by damdo** [comment]
/assign @JoelSpeed 
For a review

**Comment by damdo** [comment]
Unit tests keep flaking on that single balancing zones test, which is not related to this PR, and which is being investigated, hence overriding:
/override ci/prow/unit

**Comment by damdo** [comment]
Overriding unrelated telemetry error in aws etcd scaling, given that the rest of the tests passed.
/override ci/prow/e2e-aws-ovn-etcd-scaling

**Pull request merged**